File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Dummy node simplifies edge cases (especially when head changes after reversal)
        dummy = ListNode(0, head)
        
        # groupPrev always points to the node BEFORE the current group
        groupPrev = dummy

        while True:
            # Find the kth node from groupPrev
            # This marks the end of the current group
            kth = self._getKth(groupPrev, k)

            # If fewer than k nodes remain, stop (no reversal)
            if not kth:
                break

            # Node after the kth node (start of next group)
            groupNext = kth.next
            
            # Reverse the group:
            # prev starts from groupNext so that the reversed group connects correctly
            prev = kth.next
            curr = groupPrev.next  # first node in current group

            # Reverse nodes in the current group
            # Stop when we reach the node after the group
            while curr != groupNext:
                tmp = curr.next      # store next node
                curr.next = prev     # reverse pointer
                prev = curr          # move prev forward
                curr = tmp           # move curr forward

            # After reversal:
            # prev = new head of reversed group (which is kth)
            # groupPrev.next was the old head (now the tail)

            # Store the start of group (which becomes tail after reversal)
            tmp = groupPrev.next

            # Connect previous part to the new head (kth)
            groupPrev.next = kth

            # Move groupPrev to the end of the reversed group
            # (which is the old head)
            groupPrev = tmp
        
        return dummy.next


    def _getKth(self, curr, k):
        # Move k steps ahead from curr
        # Returns the kth node OR None if not enough nodes
        while curr and k > 0:
            curr = curr.next
            k -= 1
        return curr

File: Data Structures & Algorithms/reorder-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null)   {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode l2 = slow.next;
        slow.next = null;
        ListNode prev = null;
        while (l2 != null)    {
            ListNode temp = l2.next;
            l2.next = prev;
            prev = l2;
            l2 = temp;
        }

        ListNode l1 = head;
        l2 = prev;
        while (l2 != null)    {
            ListNode tmp1 = l1.next;
            ListNode tmp2 = l2.next;
            l1.next = l2;
            l2.next = tmp1;
            l1 = tmp1;
            l2 = tmp2;
        }
    }
}
